csi/controller_server.go

```diff
@@ -335,12 +374,176 @@ func (cs *ControllerServer) GetCapacity(context.Context, *csi.GetCapacityRequest
+	}
+
+	if backupStatus != nil && backupStatus.Progress != 100 && backupStatus.Error == "" {
+		var creationTime string
+		if snapshot != nil {
+			creationTime = snapshot.Created
+		}
+
+		rsp := createSnapshotResponse(csiVolumeName, backupStatus.Id, creationTime, existVol.Size, int(backupStatus.Progress))
+		logrus.Infof("ControllerServer CreateSnapshot rsp: %v", rsp)
+		return rsp, nil
+	}
+
+	// no existing backup and no local snapshot, create a new one
+	if snapshot == nil {
+		snapshot, err = cs.apiClient.Volume.ActionSnapshotCreate(existVol, &longhornclient.SnapshotInput{
+			Labels: csiLabels,
+			Name:   csiSnapshotName,
+		})
+
+		// failed to create snapshot, so there is no way to backup
+		if err != nil {
+			return nil, status.Error(codes.Internal, err.Error())
+		}
+	}
+
+	// create backup based on local volume snapshot
+	existVol, err = cs.apiClient.Volume.ActionSnapshotBackup(existVol, &longhornclient.SnapshotInput{
+		Labels: csiLabels,
+		Name:   csiSnapshotName,
+	})
+
+	// failed to kick off backup
+	if err != nil {
+		return nil, status.Error(codes.Internal, err.Error())
+	}
+
+	// we need to wait for backup initiation since we only know the backupID after the fact
+	backupStatus, err = cs.waitForBackupInitiation(csiVolumeName, csiSnapshotName)
+	if err != nil {
+		return nil, err
+	}
+
+	rsp := createSnapshotResponse(existVol.Name, backupStatus.Id, snapshot.Created, existVol.Size, int(backupStatus.Progress))
+	logrus.Debugf("ControllerServer CreateSnapshot rsp: %v", rsp)
+	return rsp, nil
 }
 
-func (cs *ControllerServer) DeleteSnapshot(context.Context, *csi.DeleteSnapshotRequest) (*csi.DeleteSnapshotResponse, error) {
-	return nil, status.Error(codes.Unimplemented, "")
+func createSnapshotResponse(volumeName string, backupName string, snapshotTime string, volumeSize string, progress int) *csi.CreateSnapshotResponse {
+	creationTime, err := toProtoTimestamp(snapshotTime)
+	if err != nil {
+		logrus.Errorf("Failed to parse creation time %v for backup %v", snapshotTime, backupName)
+	}
+
+	size, _ := util.ConvertSize(volumeSize)
+	size = util.RoundUpSize(size)
+	snapshotID := encodeSnapshotID(volumeName, backupName)
+	return &csi.CreateSnapshotResponse{
+		Snapshot: &csi.Snapshot{
+			SizeBytes:      size,
+			SnapshotId:     snapshotID,
+			SourceVolumeId: volumeName,
+			CreationTime:   creationTime,
+			ReadyToUse:     progress == 100,
+		},
+	}
+}
+
+// encodeSnapshotID cncodes the backup volume as part of the snapshotID
+// otherwise we need to iterate over all backup volumes,
+// when trying to find a backup for deletion or restoration
+func encodeSnapshotID(volumeName string, backupName string) string {
+	return fmt.Sprintf("backupstore/%s/%s", volumeName, backupName)
+}
+
+// decodeSnapshotID splits up the snapshotID back into it's components
+// backupType will be used once we implement the VolumeBackup crd
+func decodeSnapshotID(snapshotID string) (backupType string, volumeName string, backupName string) {
```

Comment:
```suggestion
func decodeSnapshotID(snapshotID string) (backupType, volumeName, backupName string) {
```

Reply:
Thanks I prefer to specify the types explicitly, It makes it easier to see what is what, it also makes it easier to refactor the params down the line.
What's your reasoning for suggesting this change?

Reply:
I am fine with either way. Use `xxx, xxx, xxx <type>` might be more of a Go thing.

Reply:
Np, more of a preference thing. To me, cleaning up the duplication of the same type names sometimes could help to make it a bit easier to read.

Reply:
Thanks for the discussion, I removed the duplicate type parameters.
I checked the code base and we seems to remove them in most places, it's better to have a consistent code base.
When in rome do like the romans :)

I generally prefer to be more verbose the wider the scope of the variables/parameters is.
As example using `i` as a loop variable is totally fine with me, using `i` as a function parameter is discussable